lib/validator/task/mod.rs

```diff
@@ -1155,15 +1155,446 @@ where
+            }
+        }
+        // Events should only ever be sent after committing DB txs, see
+        // https://github.com/LayerTwo-Labs/bip300301_enforcer/pull/185
+        let _send_err: Result<Option<_>, TrySendError<_>> = event_tx.try_broadcast(event);
+    }
+
+    let () = rwtxn.commit()?;
+
+    tracing::info!(
+        total_txs = total_txs,
+        total_bmm_commitments = total_bmm_commitments,
+        total_events = total_events,
+        "Synced batch of {} blocks in {}",
+        blocks.len(),
+        jiff::SignedDuration::try_from(start.elapsed()).unwrap_or_default(),
+    );
+    Ok(())
+}
+
+/// Simple LRU-style cache for temporarily storing blocks read from disk
+/// that don't match the current expected block hash.
+struct BlockCache {
+    blocks: HashMap<BlockHash, Block>,
+    insertion_order: VecDeque<BlockHash>,
+    max_size: usize,
+}
+
+impl BlockCache {
+    fn new(max_size: usize) -> Self {
+        Self {
+            blocks: HashMap::new(),
+            insertion_order: VecDeque::new(),
+            max_size,
+        }
+    }
+
+    fn insert(&mut self, block_hash: BlockHash, block: Block) {
+        // If block already exists, don't insert again
+        if self.blocks.contains_key(&block_hash) {
+            return;
+        }
+
+        // Evict oldest block if at capacity
+        while self.blocks.len() >= self.max_size {
+            if let Some(oldest_hash) = self.insertion_order.pop_front() {
+                self.blocks.remove(&oldest_hash);
+                tracing::trace!("Evicted block `{}` from cache", oldest_hash);
+            }
+        }
+
+        // Insert new block
+        self.blocks.insert(block_hash, block);
+        self.insertion_order.push_back(block_hash);
+        tracing::trace!(
+            "Cached block `{}` (cache size: {})",
+            block_hash,
+            self.blocks.len()
+        );
+    }
+
+    fn remove(&mut self, block_hash: &BlockHash) -> Option<Block> {
+        if let Some(block) = self.blocks.remove(block_hash) {
+            // Remove from insertion order
+            if let Some(pos) = self.insertion_order.iter().position(|h| h == block_hash) {
+                self.insertion_order.remove(pos);
+            }
+            tracing::trace!("Cache hit for block `{}`", block_hash,);
+            Some(block)
+        } else {
+            None
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.blocks.len()
+    }
+}
+
+/// Configuration constants for block file parsing with caching
+const MAX_BLOCK_CACHE_SIZE: usize = 5000;
+const MAX_ITERATIONS_WITHOUT_MATCH: usize = 5000;
+const BLOCKS_DIR_CONNECT_BATCH_SIZE: usize = 2000;
+
+/// Check cache for subsequent expected blocks and process them if found.
+/// This function iteratively looks for the next expected block in the cache
+/// and processes it until no more consecutive blocks are found.
+fn process_cached_blocks(
```

Comment:
Same here, IMO it would be best to clearly indicate to the caller that you are going to do a DB write by having a
`mut rwtxn: RwTxn<'a>` argument. Eg.
```rust
fn process_cached_blocks<'txn>(
     dbs: &Dbs,
     mut rwtxn: RwTxn<'a>,
     ...,
) -> Result<(), error::Sync> {
    ...
    let () = handle_block_batch(dbs, rwtxn, ...)?;
    rwtxn = dbs.write_txn()?;
    ...
}
```